### PR TITLE
backend/feature: Dynamic Fare Policy & Flow Type & Special Zone Tag

### PR DESCRIPTION
**Pull request by khuzema786**
## Type of Change

- [] Bugfix
- [x] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnConfirm.hs [hunk 1]
```diff
@@ -17,27 +17,29 @@ module Beckn.ACL.OnConfirm (buildOnConfirmMessage) where
...
 import Kernel.Prelude
 import Kernel.Types.Error
 import Kernel.Types.Id
 import Kernel.Utils.Common
 import SharedLogic.FareCalculator
+import qualified Storage.Queries.FareProduct as QFP
+import qualified Domain.Types.FareProduct as DFP
 
-buildOnConfirmMessage :: MonadFlow m => UTCTime -> DConfirm.DConfirmRes -> m OnConfirm.OnConfirmMessage
+buildOnConfirmMessage :: (MonadFlow m, EsqDBFlow m r) => UTCTime -> DConfirm.DConfirmRes -> m OnConfirm.OnConfirmMessage
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnConfirm.hs:29]: `MonadFlow m` is already included in `EsqDBFlow m r`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareParameters.hs [hunk 2]
```diff
@@ -15,13 +15,15 @@
...
 
+import qualified Domain.Types.FareProduct as DFP
 import Kernel.Prelude
 import Kernel.Types.Id
 import Kernel.Utils.Common
 import Kernel.Utils.GenericPretty
 
 data FareParameters = FareParameters
   { id :: Id FareParameters,
+    fareProductId :: Id DFP.FareProduct,
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareParameters.hs:26]: Why is this required?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+data FlowType = NORMAL_RIDE_FLOW | RIDE_OTP_FLOW deriving (Generic, Eq, Show, Read, FromJSON, ToJSON)
+
+data FareProduct = FareProduct
+  { id :: Id FareProduct,
+    merchantId :: Id DM.Merchant,
+    vehicleVariant :: Vehicle.Variant,
+    farePolicyType :: FarePolicyType,
+    flowType :: FlowType
+  }
+  deriving (Generic, Show, FromJSON, ToJSON)
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:33]: Don't make  FromJSON, ToJSON instances for domain types

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SpecialZoneCategoryPriority.hs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import Kernel.Types.Id (Id)
+
+data SpecialZoneCategoryPriority = SpecialZoneCategoryPriority
+  { id :: Id SpecialZoneCategoryPriority,
+    category :: Text,
+    merchantId :: Id DM.Merchant,
+    pickupPriority :: Int,
+    dropPriority :: Int
+  }
+  deriving (Generic, Show, FromJSON, ToJSON)
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/SpecialZoneCategoryPriority.hs:28]: Same

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import qualified Domain.Types.Merchant as DM
+import qualified Domain.Types.Vehicle as Vehicle
+import Kernel.Prelude
+import Kernel.Types.Id (Id)
+
+data FarePolicyType = NORMAL | SLAB deriving (Generic, Eq, Show, Read, FromJSON, ToJSON)
+
+data FlowType = NORMAL_RIDE_FLOW | RIDE_OTP_FLOW deriving (Generic, Eq, Show, Read, FromJSON, ToJSON)
+
+data FareProduct = FareProduct
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:26]: Could you describe why we need this table?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+module Domain.Types.FareProduct where
+
+import qualified Domain.Types.Merchant as DM
+import qualified Domain.Types.Vehicle as Vehicle
+import Kernel.Prelude
+import Kernel.Types.Id (Id)
+
+data FarePolicyType = NORMAL | SLAB deriving (Generic, Eq, Show, Read, FromJSON, ToJSON)
+
+data FlowType = NORMAL_RIDE_FLOW | RIDE_OTP_FLOW deriving (Generic, Eq, Show, Read, FromJSON, ToJSON)
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:24]: What is this naming? Don't we have OTP in all rides? What is considered "Normal" here then?

**Review by DonDokidon — CHANGES_REQUESTED** [review]

**Comment by khuzema786** [comment]
Outdated, raised a new PR https://github.com/nammayatri/nammayatri/pull/978

**Pull request closed without merging**